Sporadic test failure OperationProcessorTests.testWithDataCorruptionFailures

**Problem description**
We are getting sporadic test failures in Jenkins related to `OperationProcessorTests.testWithDataCorruptionFailures`:
```
java.lang.AssertionError: Unexpected truncation marker for Operation SeqNo 100 when it spans multiple DataFrames. expected:<Sequence = 0> but was:<null>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at io.pravega.segmentstore.server.logs.OperationProcessorTests.performLogOperationChecks(OperationProcessorTests.java:541)
	at io.pravega.segmentstore.server.logs.OperationProcessorTests.testWithDataCorruptionFailures(OperationProcessorTests.java:439)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

**Problem location**
`OperationProcessorTests.testWithDataCorruptionFailures`

**Suggestions for an improvement**
Fix the test.

--- a/segmentstore/server/src/test/java/io/pravega/segmentstore/server/logs/OperationProcessorTests.java
+++ b/segmentstore/server/src/test/java/io/pravega/segmentstore/server/logs/OperationProcessorTests.java
@@ -546,22 +546,26 @@ public class OperationProcessorTests extends OperationLogTestBase {
             }
 
             // Check truncation markers if this is the last Operation to be written.
-            LogAddress dataFrameAddress = truncationMarkers.getClosestTruncationMarker(expectedOp.getSequenceNumber());
             if (dataFrameRecord.getLastFullDataFrameAddress() != null
                     && dataFrameRecord.getLastFullDataFrameAddress().getSequence() != dataFrameRecord.getLastUsedDataFrameAddress().getSequence()) {
                 // This operation spans multiple DataFrames. The TruncationMarker should be set on the last DataFrame
                 // that ends with a part of it.
-                Assert.assertEquals("Unexpected truncation marker for Operation SeqNo " + expectedOp.getSequenceNumber() + " when it spans multiple DataFrames.",
-                        dataFrameRecord.getLastFullDataFrameAddress(), dataFrameAddress);
+                AssertExtensions.assertEventuallyEquals(
+                        "Unexpected truncation marker for Operation SeqNo " + expectedOp.getSequenceNumber() + " when it spans multiple DataFrames.",
+                        dataFrameRecord.getLastFullDataFrameAddress(),
+                        () -> truncationMarkers.getClosestTruncationMarker(expectedOp.getSequenceNumber()), 100, TIMEOUT.toMillis());
             } else if (dataFrameRecord.isLastFrameEntry()) {
                 // The operation was the last one in the frame. This is a Truncation Marker.
-                Assert.assertEquals("Unexpected truncation marker for Operation SeqNo " + expectedOp.getSequenceNumber() + " when it is the last entry in a DataFrame.",
-                        dataFrameRecord.getLastUsedDataFrameAddress(), dataFrameAddress);
+                AssertExtensions.assertEventuallyEquals(
+                        "Unexpected truncation marker for Operation SeqNo " + expectedOp.getSequenceNumber() + " when it is the last entry in a DataFrame.",
+                        dataFrameRecord.getLastUsedDataFrameAddress(),
+                        () -> truncationMarkers.getClosestTruncationMarker(expectedOp.getSequenceNumber()), 100, TIMEOUT.toMillis());
             } else {
                 // The operation is not the last in the frame, and it doesn't span multiple frames either.
                 // There could be data after it that is not safe to truncate. The correct Truncation Marker is the
                 // same as the one for the previous operation.
                 LogAddress expectedTruncationMarker = truncationMarkers.getClosestTruncationMarker(expectedOp.getSequenceNumber() - 1);
+                LogAddress dataFrameAddress = truncationMarkers.getClosestTruncationMarker(expectedOp.getSequenceNumber());
                 Assert.assertEquals("Unexpected truncation marker for Operation SeqNo " + expectedOp.getSequenceNumber() + " when it is in the middle of a DataFrame.",
                         expectedTruncationMarker, dataFrameAddress);
             }

--- a/test/testcommon/src/main/java/io/pravega/test/common/AssertExtensions.java
+++ b/test/testcommon/src/main/java/io/pravega/test/common/AssertExtensions.java
@@ -67,6 +67,28 @@ public class AssertExtensions {
     }
 
     /**
+     * Invokes `eval` in a loop over and over (with a small sleep) and asserts that the value eventually reaches `expected`.
+     *
+     * @param <T>                 The type of the value to compare.
+     * @param message             The message ot include.
+     * @param expected            The expected return value.
+     * @param eval                The function to test
+     * @param checkIntervalMillis The number of milliseconds to wait between two checks.
+     * @param timeoutMillis       The timeout in milliseconds after which an assertion error should be thrown.
+     * @throws Exception If the is an assertion error, and exception from `eval`, or the thread is interrupted.
+     */
+    public static <T> void assertEventuallyEquals(String message, T expected, Callable<T> eval, int checkIntervalMillis, long timeoutMillis) throws Exception {
+        long remainingMillis = timeoutMillis;
+        while (remainingMillis > 0) {
+            if ((expected == null && eval.call() == null) || (expected != null && expected.equals(eval.call()))) {
+                return;
+            }
+            Thread.sleep(checkIntervalMillis);
+        }
+        assertEquals(message, expected, eval.call());
+    }
+
+    /**
      * Asserts that an exception of the Type provided is thrown.
      *
      * @param run  The Runnable to execute.

